File: backend/src/ontoforge_server/adapters/neo4j/ddl.py
"""Neo4j naming conventions, vector-index DDL, and index-metadata limits.

Adapter-private. Labels, index names, and the PascalCase/UPPER_SNAKE_CASE
conventions are implementation details of this adapter and must not leak
through the persistence port. The reserved-key sets below are the one
exception: they cross the port as plain type keys (never as labels), so the
modeling service can reject colliding keys without knowing why they collide.
"""

import logging
import re
from typing import Any

from neo4j import AsyncDriver

from ontoforge_server.core.exceptions import ValidationError

logger = logging.getLogger(__name__)

MAX_VECTOR_FILTER_VALUE_BYTES = 32766
ENTITY_VECTOR_INDEX_NAME = "entity_embedding"

#: Node labels this adapter uses to store schema objects.
SCHEMA_LABELS = frozenset(
    {
        "Ontology",
        "EntityType",
        "RelationType",
        "PropertyDefinition",
        "AiAgentConfig",
        "SavedQuery",
    }
)

#: Relationship types this adapter uses to connect schema objects.
SCHEMA_RELATIONSHIP_TYPES = frozenset(
    {
        "INCLUDES_TYPE",
        "HAS_PROPERTY",
        "RELATES_FROM",
        "RELATES_TO",
        "HAS_AI_AGENT",
        "HAS_SAVED_QUERY",
    }
)

# The internal names `_Entity`, `_Chunk` and `_HAS_CHUNK` need no reserved
# key: the type key pattern forbids a leading underscore, so no key can
# convert to them.


def _to_pascal_case(key: str) -> str:
    """Convert a snake_case key to PascalCase."""
    return "".join(segment.capitalize() for segment in key.split("_"))


def _to_snake_case(pascal: str) -> str:
    """Convert a PascalCase label to the snake_case key that produces it."""
    return re.sub(r"(?<!^)(?=[A-Z])", "_", pascal).lower()


def reserved_entity_type_keys() -> frozenset[str]:
    """Entity type keys whose physical label would collide with a schema label."""
    return frozenset(_to_snake_case(label) for label in SCHEMA_LABELS)


def reserved_relation_type_keys() -> frozenset[str]:
    """Relation type keys whose physical type would collide with a schema relationship."""
    return frozenset(rel_type.lower() for rel_type in SCHEMA_RELATIONSHIP_TYPES)


def document_virtual_label(entity_type_key: str, property_key: str) -> str:
    """Virtual chunk label for a document property. E.g. ('person', 'bio') -> 'PersonDocumentBio'."""
    return f"{_to_pascal_case(entity_type_key)}Document{_to_pascal_case(property_key)}"


def document_index_name(entity_type_key: str, property_key: str) -> str:
    """Vector index name for a document property's chunks."""
    return f"{entity_type_key}_document_{property_key}_embedding"


def validate_vector_indexed_properties(
    entity_type_key: str,
    properties: dict[str, Any],
    filter_properties: list[str],
    entity_id: str | None = None,
) -> None:
    """Reject string values too large for vector-index filter metadata."""
    for property_key in filter_properties:
        value = properties.get(property_key)
        if value is None or not isinstance(value, str):
            continue
        value_bytes = len(value.encode("utf-8"))
        if value_bytes <= MAX_VECTOR_FILTER_VALUE_BYTES:
            continue

        entity_ref = f" on entity '{entity_id}'" if entity_id else ""
        raise ValidationError(
            f"Property '{property_key}'{entity_ref} is too large for semantic indexing "
            f"on type '{entity_type_key}' ({value_bytes} bytes > "
            f"{MAX_VECTOR_FILTER_VALUE_BYTES} bytes)",
            details={
                "fields": {
                    property_key: (
                        "Value exceeds the indexed property size limit "
                        f"({value_bytes} bytes > {MAX_VECTOR_FILTER_VALUE_BYTES} bytes)"
                    )
                }
            },
        )


async def _validate_existing_vector_indexed_properties(
    driver: AsyncDriver,
    pascal_label: str,
    entity_type_key: str,
    filter_properties: list[str],
) -> None:
    if not filter_properties:
        return

    async with driver.session() as session:
        result = await session.run(
            f"MATCH (n:{pascal_label}) RETURN n._id AS entity_id, n {{.*}} AS properties"
        )
        async for record in result:
            validate_vector_indexed_properties(
                entity_type_key,
                record["properties"] or {},
                filter_properties,
                entity_id=record["entity_id"],
            )


async def _drop_failed_index_if_exists(driver: AsyncDriver, index_name: str) -> None:
    async with driver.session() as session:
        result = await session.run(
            "SHOW INDEXES YIELD name, state WHERE name = $name RETURN state",
            name=index_name,
        )
        record = await result.single()
        if record and record["state"] == "FAILED":
            await session.run(f"DROP INDEX {index_name} IF EXISTS")
            logger.warning("Dropped failed vector index before recreate: %s", index_name)


async def ensure_vector_indexes(driver: AsyncDriver, dimensions: int) -> None:
    """Create vector indexes for all existing entity types (IF NOT EXISTS).

    New indexes include a WITH clause listing all current properties for
    in-index filtering (Neo4j 2026+ SEARCH clause). Existing indexes are
    left untouched.
    """
    async with driver.session() as session:
        result = await session.run(
            """
            MATCH (et:EntityType)
            OPTIONAL MATCH (et)-[:HAS_PROPERTY]->(p:PropertyDefinition)
            WHERE p.dataType <> 'document'
            RETURN et.key AS key, collect(p.key) AS property_keys
            """
        )
        entity_types = [
            {"key": record["key"], "property_keys": record["property_keys"]}
            async for record in result
        ]

    for et in entity_types:
        await create_vector_index(
            driver, et["key"], dimensions, filter_properties=et["property_keys"]
        )

    # Chunk vector indexes for document properties (one per virtual type)
    async with driver.session() as session:
        result = await session.run(
            """
            MATCH (et:EntityType)-[:HAS_PROPERTY]->(p:PropertyDefinition {dataType: 'document'})
            RETURN et.key AS entity_type_key, p.key AS property_key
            """
        )
        document_properties = [
            (record["entity_type_key"], record["property_key"])
            async for record in result
        ]

    for entity_type_key, property_key in document_properties:
        await create_document_vector_index(
            driver, entity_type_key, property_key, dimensions
        )

    # Cross-type entity vector index (semantic search across all types)
    await ensure_entity_vector_index(driver, dimensions)

    # Saved query vector index (for semantic search over descriptions)
    await ensure_saved_query_vector_index(driver, dimensions)


async def ensure_entity_vector_index(driver: AsyncDriver, dimensions: int) -> None:
    """Create the cross-type vector index on the shared _Entity label (IF NOT EXISTS).

    Indexes _embedding across all entity instances so semantic search can run
    over every entity type in a single query. Type/scope filtering happens in
    the service layer, so no in-index filter properties are needed.
    """
    await _drop_failed_index_if_exists(driver, ENTITY_VECTOR_INDEX_NAME)
    query = (
        f"CREATE VECTOR INDEX {ENTITY_VECTOR_INDEX_NAME} IF NOT EXISTS "
        "FOR (n:_Entity) ON (n._embedding) "
        f"OPTIONS {{indexConfig: {{`vector.dimensions`: {dimensions}, "
        f"`vector.similarity_function`: 'cosine'}}}}"
    )
    async with driver.session() as session:
        await session.run(query)
    logger.info("Vector index ensured: %s", ENTITY_VECTOR_INDEX_NAME)


async def ensure_saved_query_vector_index(
    driver: AsyncDriver, dimensions: int
) -> None:
    """Create the vector index for SavedQuery descriptions (IF NOT EXISTS).

    Uses _ontologyKey as an in-index filter property so that semantic search
    can be scoped to a single ontology in one query.
    """
    query = (
        "CREATE VECTOR INDEX saved_query_embedding IF NOT EXISTS "
        "FOR (sq:SavedQuery) ON (sq._embedding) "
        "WITH [sq._ontologyKey] "
        f"OPTIONS {{indexConfig: {{`vector.dimensions`: {dimensions}, "
        f"`vector.similarity_function`: 'cosine'}}}}"
    )
    async with driver.session() as session:
        await session.run(query)
    logger.info("Vector index ensured: saved_query_embedding")


async def create_vector_index(
    driver: AsyncDriver,
    entity_type_key: str,
    dimensions: int,
    filter_properties: list[str] | None = None,
) -> None:
    """Create a vector index for the given entity type label.

    When *filter_properties* is provided, the index is created with a WITH
    clause so that those properties are stored alongside vectors for in-index
    filtering (Neo4j 2026+ SEARCH clause).
    """
    pascal_label = _to_pascal_case(entity_type_key)
    index_name = f"{entity_type_key}_embedding"
    selected_properties = [p for p in (filter_properties or []) if p]
    await _validate_existing_vector_indexed_properties(
        driver, pascal_label, entity_type_key, selected_properties
    )
    await _drop_failed_index_if_exists(driver, index_name)
    with_clause = ""
    if selected_properties:
        props = ", ".join(f"n.{p}" for p in selected_properties)
        with_clause = f"WITH [{props}] "
    query = (
        f"CREATE VECTOR INDEX {index_name} IF NOT EXISTS "
        f"FOR (n:{pascal_label}) ON (n._embedding) "
        f"{with_clause}"
        f"OPTIONS {{indexConfig: {{`vector.dimensions`: {dimensions}, "
        f"`vector.similarity_function`: 'cosine'}}}}"
    )
    async with driver.session() as session:
        await session.run(query)
    logger.info("Vector index ensured: %s", index_name)


async def create_document_vector_index(
    driver: AsyncDriver,
    entity_type_key: str,
    property_key: str,
    dimensions: int,
) -> None:
    """Create the vector index for a document property's chunk nodes."""
    index_name = document_index_name(entity_type_key, property_key)
    virtual_label = document_virtual_label(entity_type_key, property_key)
    await _drop_failed_index_if_exists(driver, index_name)
    query = (
        f"CREATE VECTOR INDEX {index_name} IF NOT EXISTS "
        f"FOR (c:{virtual_label}) ON (c._embedding) "
        f"OPTIONS {{indexConfig: {{`vector.dimensions`: {dimensions}, "
        f"`vector.similarity_function`: 'cosine'}}}}"
    )
    async with driver.session() as session:
        await session.run(query)
    logger.info("Vector index ensured: %s", index_name)


async def drop_document_vector_index(
    driver: AsyncDriver, entity_type_key: str, property_key: str
) -> None:
    """Drop the vector index for a document property's chunk nodes."""
    index_name = document_index_name(entity_type_key, property_key)
    async with driver.session() as session:
        await session.run(f"DROP INDEX {index_name} IF EXISTS")
    logger.info("Vector index dropped: %s", index_name)


async def drop_vector_index(driver: AsyncDriver, entity_type_key: str) -> None:
    """Drop the vector index for the given entity type."""
    index_name = f"{entity_type_key}_embedding"
    async with driver.session() as session:
        await session.run(f"DROP INDEX {index_name} IF EXISTS")
    logger.info("Vector index dropped: %s", index_name)


async def rebuild_vector_index(
    driver: AsyncDriver, entity_type_key: str, dimensions: int
) -> None:
    """Drop and recreate the vector index with current properties.

    Called when properties are added or removed from an entity type so that
    the in-index filter properties stay in sync with the schema.
    """
    async with driver.session() as session:
        result = await session.run(
            """
            MATCH (et:EntityType {key: $key})
            OPTIONAL MATCH (et)-[:HAS_PROPERTY]->(p:PropertyDefinition)
            WHERE p.dataType <> 'document'
            RETURN collect(p.key) AS property_keys
            """,
            key=entity_type_key,
        )
        record = await result.single()
        property_keys = record["property_keys"] if record else []

    await _validate_existing_vector_indexed_properties(
        driver, _to_pascal_case(entity_type_key), entity_type_key, property_keys
    )
    await drop_vector_index(driver, entity_type_key)
    await create_vector_index(
        driver, entity_type_key, dimensions, filter_properties=property_keys
    )

File: server/src/adapters/neo4j/ddl.ts
/**
 * Neo4j naming conventions, vector-index DDL, and index-metadata limits.
 *
 * Adapter-private. Labels, index names, and the PascalCase/UPPER_SNAKE_CASE
 * conventions are implementation details of this adapter and must not leak
 * through the persistence port. The reserved-key sets below are the one
 * exception: they cross the port as plain type keys (never as labels), so
 * the modeling service can reject colliding keys without knowing why they
 * collide.
 */

import type { Driver } from "neo4j-driver";

import { ValidationError } from "../../core/exceptions.js";
import {
  ALL_ENTITY_TYPES_SCOPE,
  documentPropertyScope,
  entityTypeScope,
  reportWidthMismatch,
  reportWidthRecreate,
  SAVED_QUERY_SCOPE,
} from "../../core/vectorDrift.js";
import { runSession } from "./errors.js";

/** Node labels this adapter uses to store schema objects. */
export const SCHEMA_LABELS: ReadonlySet<string> = new Set([
  "Ontology",
  "EntityType",
  "RelationType",
  "PropertyDefinition",
  "AiAgentConfig",
  "SavedQuery",
]);

/** Relationship types this adapter uses to connect schema objects. */
export const SCHEMA_RELATIONSHIP_TYPES: ReadonlySet<string> = new Set([
  "INCLUDES_TYPE",
  "HAS_PROPERTY",
  "RELATES_FROM",
  "RELATES_TO",
  "HAS_AI_AGENT",
  "HAS_SAVED_QUERY",
]);

// The internal names `_Entity`, `_Chunk` and `_HAS_CHUNK` need no reserved
// key: the type key pattern forbids a leading underscore, so no key can
// convert to them.

/** Convert a snake_case key to its PascalCase label. */
export function toPascalCase(key: string): string {
  return key
    .split("_")
    .map((segment) => segment.charAt(0).toUpperCase() + segment.slice(1))
    .join("");
}

/** Convert a PascalCase label to the snake_case key that produces it. */
export function toSnakeCase(pascal: string): string {
  return pascal.replace(/(?<!^)(?=[A-Z])/g, "_").toLowerCase();
}

/** Convert a snake_case key to its UPPER_SNAKE_CASE relationship type. */
export function toUpperSnakeCase(key: string): string {
  return key.toUpperCase();
}

/** Virtual chunk label for a document property.
 * E.g. ('person', 'bio') -> 'PersonDocumentBio'. */
export function documentVirtualLabel(entityTypeKey: string, propertyKey: string): string {
  return `${toPascalCase(entityTypeKey)}Document${toPascalCase(propertyKey)}`;
}

/** Vector index name for a document property's chunks. */
export function documentIndexName(entityTypeKey: string, propertyKey: string): string {
  return `${entityTypeKey}_document_${propertyKey}_embedding`;
}

/**
 * Entity type keys whose physical label would collide with a schema label.
 *
 * Derived — not copied — by applying the inverse naming transformation to
 * the labels this adapter actually uses, so the set can never drift from
 * the physical naming that makes the collision real.
 */
export function reservedEntityTypeKeys(): ReadonlySet<string> {
  return new Set([...SCHEMA_LABELS].map(toSnakeCase));
}

/**
 * Relation type keys whose physical relationship type would collide with a
 * schema relationship type. Derived from the physical names, as above.
 */
export function reservedRelationTypeKeys(): ReadonlySet<string> {
  return new Set([...SCHEMA_RELATIONSHIP_TYPES].map((relType) => relType.toLowerCase()));
}

/**
 * Constraints and indexes created unconditionally at startup.
 */
export const CONSTRAINTS: readonly string[] = [
  "CREATE CONSTRAINT ontology_id_unique IF NOT EXISTS FOR (o:Ontology) REQUIRE o.ontologyId IS UNIQUE",
  "CREATE CONSTRAINT ontology_key_unique IF NOT EXISTS FOR (o:Ontology) REQUIRE o.key IS UNIQUE",
  "CREATE CONSTRAINT ontology_name_unique IF NOT EXISTS FOR (o:Ontology) REQUIRE o.name IS UNIQUE",
  "CREATE CONSTRAINT entity_type_id_unique IF NOT EXISTS FOR (et:EntityType) REQUIRE et.entityTypeId IS UNIQUE",
  "CREATE CONSTRAINT entity_type_key_unique IF NOT EXISTS FOR (et:EntityType) REQUIRE et.key IS UNIQUE",
  "CREATE CONSTRAINT relation_type_id_unique IF NOT EXISTS FOR (rt:RelationType) REQUIRE rt.relationTypeId IS UNIQUE",
  "CREATE CONSTRAINT relation_type_key_unique IF NOT EXISTS FOR (rt:RelationType) REQUIRE rt.key IS UNIQUE",
  "CREATE CONSTRAINT property_id_unique IF NOT EXISTS FOR (pd:PropertyDefinition) REQUIRE pd.propertyId IS UNIQUE",
  "CREATE CONSTRAINT entity_instance_id_unique IF NOT EXISTS FOR (n:_Entity) REQUIRE n._id IS UNIQUE",
  "CREATE INDEX entity_type_key_index IF NOT EXISTS FOR (n:_Entity) ON (n._entityTypeKey)",
  "CREATE CONSTRAINT agent_config_id_unique IF NOT EXISTS FOR (ac:AiAgentConfig) REQUIRE ac.agentConfigId IS UNIQUE",
  "CREATE CONSTRAINT saved_query_id_unique IF NOT EXISTS FOR (sq:SavedQuery) REQUIRE sq.savedQueryId IS UNIQUE",
];

// ---------------------------------------------------------------------------
// Vector-index metadata limits
// ---------------------------------------------------------------------------

/** The engine's indexed-property size ceiling: a per-type vector index
 * stores property values as filter metadata, so indexed string values may
 * not exceed it (`docs/storage-adapters.md`, engine constraints). */
export const MAX_VECTOR_FILTER_VALUE_BYTES = 32766;

export const ENTITY_VECTOR_INDEX_NAME = "entity_embedding";

/**
 * Reject string values too large for vector-index filter metadata.
 *
 * Raised as a domain validation error naming the property, never the
 * engine — enforced only when an embedding provider is configured, since
 * the constraint exists to protect the index.
 */
export function validateVectorIndexedProperties(
  entityTypeKey: string,
  properties: Record<string, unknown>,
  filterProperties: string[],
  entityId: string | null = null,
): void {
  for (const propertyKey of filterProperties) {
    const value = properties[propertyKey];
    if (value === null || value === undefined || typeof value !== "string") {
      continue;
    }
    const valueBytes = Buffer.byteLength(value, "utf8");
    if (valueBytes <= MAX_VECTOR_FILTER_VALUE_BYTES) {
      continue;
    }

    const entityRef = entityId ? ` on entity '${entityId}'` : "";
    throw new ValidationError(
      `Property '${propertyKey}'${entityRef} is too large for semantic indexing ` +
        `on type '${entityTypeKey}' (${valueBytes} bytes > ` +
        `${MAX_VECTOR_FILTER_VALUE_BYTES} bytes)`,
      {
        fields: {
          [propertyKey]:
            "Value exceeds the indexed property size limit " +
            `(${valueBytes} bytes > ${MAX_VECTOR_FILTER_VALUE_BYTES} bytes)`,
        },
      },
    );
  }
}

async function validateExistingVectorIndexedProperties(
  driver: Driver,
  pascalLabel: string,
  entityTypeKey: string,
  filterProperties: string[],
): Promise<void> {
  if (filterProperties.length === 0) {
    return;
  }

  await runSession(driver, async (session) => {
    const result = await session.run(
      `MATCH (n:${pascalLabel}) RETURN n._id AS entity_id, n {.*} AS properties`,
    );
    for (const record of result.records) {
      validateVectorIndexedProperties(
        entityTypeKey,
        (record.get("properties") as Record<string, unknown>) ?? {},
        filterProperties,
        record.get("entity_id") as string,
      );
    }
  });
}

// ---------------------------------------------------------------------------
// Vector-index lifecycle
// ---------------------------------------------------------------------------

/** A failed index is silently useless: a create-if-absent would skip over
 * it forever, so it is dropped before recreate. */
async function dropFailedIndexIfExists(driver: Driver, indexName: string): Promise<void> {
  await runSession(driver, async (session) => {
    const result = await session.run(
      "SHOW INDEXES YIELD name, state WHERE name = $name RETURN state",
      { name: indexName },
    );
    const record = result.records[0];
    if (record !== undefined && record.get("state") === "FAILED") {
      await session.run(`DROP INDEX ${indexName} IF EXISTS`);
      console.warn(`Dropped failed vector index before recreate: ${indexName}`);
    }
  });
}

/** The vector width an existing index is configured for, or null if absent. */
export async function existingVectorIndexDimensions(
  driver: Driver,
  indexName: string,
): Promise<number | null> {
  return runSession(driver, async (session) => {
    const result = await session.run(
      "SHOW VECTOR INDEXES YIELD name, options WHERE name = $name RETURN options",
      { name: indexName },
    );
    const record = result.records[0];
    if (record === undefined) {
      return null;
    }
    const options = (record.get("options") as Record<string, unknown> | null) ?? {};
    const indexConfig = (options.indexConfig as Record<string, unknown> | undefined) ?? {};
    const configured = indexConfig["vector.dimensions"];
    return configured === null || configured === undefined ? null : Number(configured);
  });
}

/**
 * Handle an existing index whose width no longer matches the model.
 *
 * A vector index fixes its width at creation, and `CREATE ... IF NOT
 * EXISTS` is a no-op against one that already exists — so changing the
 * embedding model leaves an index that rejects every vector the new model
 * produces. Nothing else notices: the index is ONLINE, so the failed-index
 * check above does not see it, and the symptom only appears later, as a
 * storage failure on the first semantic search.
 *
 * On startup this only warns: dropping an index destroys the vectors it
 * holds, and that is the operator's call. The rebuild path passes
 * `recreateOnMismatch`, because there the drop is followed immediately by
 * regeneration at the new width — the operator asked for exactly that.
 *
 * Only the detection is here. What the operator is told — the wording
 * and the API-scope vocabulary every backend shares — is
 * `core/vectorDrift.ts`.
 */
export async function reconcileIndexDimensions(
  driver: Driver,
  indexName: string,
  describes: string,
  dimensions: number,
  recreateOnMismatch: boolean,
): Promise<void> {
  const existing = await existingVectorIndexDimensions(driver, indexName);
  if (existing === null || existing === dimensions) {
    return;
  }

  if (!recreateOnMismatch) {
    reportWidthMismatch(describes, existing, dimensions);
    return;
  }

  await runSession(driver, async (session) => {
    await session.run(`DROP INDEX ${indexName} IF EXISTS`);
  });
  reportWidthRecreate(describes, existing, dimensions);
}

/**
 * Create vector indexes for all existing entity types (IF NOT EXISTS).
 *
 * New indexes include a WITH clause listing all current non-document
 * properties for in-index filtering. Existing indexes are left untouched
 * unless their width no longer matches `dimensions`, in which case they
 * are reported — or, with `recreateOnMismatch`, dropped and recreated.
 * See `reconcileIndexDimensions`.
 */
export async function ensureVectorIndexes(
  driver: Driver,
  dimensions: number,
  recreateOnMismatch = false,
): Promise<void> {
  const entityTypes = await runSession(driver, async (session) => {
    const result = await session.run(
      `
      MATCH (et:EntityType)
      OPTIONAL MATCH (et)-[:HAS_PROPERTY]->(p:PropertyDefinition)
      WHERE p.dataType <> 'document'
      RETURN et.key AS key, collect(p.key) AS property_keys
      `,
    );
    return result.records.map((record) => ({
      key: record.get("key") as string,
      propertyKeys: record.get("property_keys") as string[],
    }));
  });

  for (const et of entityTypes) {
    await createVectorIndex(driver, et.key, dimensions, et.propertyKeys, recreateOnMismatch);
  }

  // Chunk vector indexes for document properties (one per virtual type).
  const documentProperties = await runSession(driver, async (session) => {
    const result = await session.run(
      `
      MATCH (et:EntityType)-[:HAS_PROPERTY]->(p:PropertyDefinition {dataType: 'document'})
      RETURN et.key AS entity_type_key, p.key AS property_key
      `,
    );
    return result.records.map((record) => ({
      entityTypeKey: record.get("entity_type_key") as string,
      propertyKey: record.get("property_key") as string,
    }));
  });

  for (const { entityTypeKey, propertyKey } of documentProperties) {
    await createDocumentVectorIndex(
      driver,
      entityTypeKey,
      propertyKey,
      dimensions,
      recreateOnMismatch,
    );
  }

  // Cross-type entity vector index (semantic search across all types).
  await ensureEntityVectorIndex(driver, dimensions, recreateOnMismatch);

  // Saved-query vector index (semantic search over descriptions).
  await ensureSavedQueryVectorIndex(driver, dimensions, recreateOnMismatch);
}

/**
 * Create the cross-type vector index on the shared `_Entity` label
 * (IF NOT EXISTS). Type/scope filtering happens in the service layer, so
 * no in-index filter properties are needed.
 */
export async function ensureEntityVectorIndex(
  driver: Driver,
  dimensions: number,
  recreateOnMismatch = false,
): Promise<void> {
  await dropFailedIndexIfExists(driver, ENTITY_VECTOR_INDEX_NAME);
  await reconcileIndexDimensions(
    driver,
    ENTITY_VECTOR_INDEX_NAME,
    ALL_ENTITY_TYPES_SCOPE,
    dimensions,
    recreateOnMismatch,
  );
  const query =
    `CREATE VECTOR INDEX ${ENTITY_VECTOR_INDEX_NAME} IF NOT EXISTS ` +
    "FOR (n:_Entity) ON (n._embedding) " +
    `OPTIONS {indexConfig: {\`vector.dimensions\`: ${dimensions}, ` +
    "`vector.similarity_function`: 'cosine'}}";
  await runSession(driver, async (session) => {
    await session.run(query);
  });
  console.info(`Vector index ensured: ${ENTITY_VECTOR_INDEX_NAME}`);
}

/**
 * Create the vector index for SavedQuery descriptions (IF NOT EXISTS).
 * `_ontologyKey` is an in-index filter property so a description search
 * can be scoped to one ontology in a single query.
 */
export async function ensureSavedQueryVectorIndex(
  driver: Driver,
  dimensions: number,
  recreateOnMismatch = false,
): Promise<void> {
  await reconcileIndexDimensions(
    driver,
    "saved_query_embedding",
    SAVED_QUERY_SCOPE,
    dimensions,
    recreateOnMismatch,
  );
  const query =
    "CREATE VECTOR INDEX saved_query_embedding IF NOT EXISTS " +
    "FOR (sq:SavedQuery) ON (sq._embedding) " +
    "WITH [sq._ontologyKey] " +
    `OPTIONS {indexConfig: {\`vector.dimensions\`: ${dimensions}, ` +
    "`vector.similarity_function`: 'cosine'}}";
  await runSession(driver, async (session) => {
    await session.run(query);
  });
  console.info("Vector index ensured: saved_query_embedding");
}

/**
 * Create a vector index for the given entity type label. When
 * `filterProperties` is provided, the index is created with a WITH clause
 * so those properties are stored alongside vectors for in-index filtering.
 */
export async function createVectorIndex(
  driver: Driver,
  entityTypeKey: string,
  dimensions: number,
  filterProperties: string[] | null = null,
  recreateOnMismatch = false,
): Promise<void> {
  const pascalLabel = toPascalCase(entityTypeKey);
  const indexName = `${entityTypeKey}_embedding`;
  const selectedProperties = (filterProperties ?? []).filter((p) => p);
  await validateExistingVectorIndexedProperties(
    driver,
    pascalLabel,
    entityTypeKey,
    selectedProperties,
  );
  await dropFailedIndexIfExists(driver, indexName);
  await reconcileIndexDimensions(
    driver,
    indexName,
    entityTypeScope(entityTypeKey),
    dimensions,
    recreateOnMismatch,
  );
  let withClause = "";
  if (selectedProperties.length > 0) {
    const props = selectedProperties.map((p) => `n.${p}`).join(", ");
    withClause = `WITH [${props}] `;
  }
  const query =
    `CREATE VECTOR INDEX ${indexName} IF NOT EXISTS ` +
    `FOR (n:${pascalLabel}) ON (n._embedding) ` +
    withClause +
    `OPTIONS {indexConfig: {\`vector.dimensions\`: ${dimensions}, ` +
    "`vector.similarity_function`: 'cosine'}}";
  await runSession(driver, async (session) => {
    await session.run(query);
  });
  console.info(`Vector index ensured: ${indexName}`);
}

/** Create the vector index for a document property's chunk nodes. */
export async function createDocumentVectorIndex(
  driver: Driver,
  entityTypeKey: string,
  propertyKey: string,
  dimensions: number,
  recreateOnMismatch = false,
): Promise<void> {
  const indexName = documentIndexName(entityTypeKey, propertyKey);
  const virtualLabel = documentVirtualLabel(entityTypeKey, propertyKey);
  await dropFailedIndexIfExists(driver, indexName);
  await reconcileIndexDimensions(
    driver,
    indexName,
    documentPropertyScope(entityTypeKey, propertyKey),
    dimensions,
    recreateOnMismatch,
  );
  const query =
    `CREATE VECTOR INDEX ${indexName} IF NOT EXISTS ` +
    `FOR (c:${virtualLabel}) ON (c._embedding) ` +
    `OPTIONS {indexConfig: {\`vector.dimensions\`: ${dimensions}, ` +
    "`vector.similarity_function`: 'cosine'}}";
  await runSession(driver, async (session) => {
    await session.run(query);
  });
  console.info(`Vector index ensured: ${indexName}`);
}

/** Drop the vector index for a document property's chunk nodes. */
export async function dropDocumentVectorIndex(
  driver: Driver,
  entityTypeKey: string,
  propertyKey: string,
): Promise<void> {
  const indexName = documentIndexName(entityTypeKey, propertyKey);
  await runSession(driver, async (session) => {
    await session.run(`DROP INDEX ${indexName} IF EXISTS`);
  });
  console.info(`Vector index dropped: ${indexName}`);
}

/** Drop the vector index for the given entity type. */
export async function dropVectorIndex(driver: Driver, entityTypeKey: string): Promise<void> {
  const indexName = `${entityTypeKey}_embedding`;
  await runSession(driver, async (session) => {
    await session.run(`DROP INDEX ${indexName} IF EXISTS`);
  });
  console.info(`Vector index dropped: ${indexName}`);
}

/**
 * Drop and recreate the vector index with current properties. Called when
 * properties are added or removed from an entity type so that the
 * in-index filter properties stay in sync with the schema.
 */
export async function rebuildVectorIndex(
  driver: Driver,
  entityTypeKey: string,
  dimensions: number,
): Promise<void> {
  const propertyKeys = await runSession(driver, async (session) => {
    const result = await session.run(
      `
      MATCH (et:EntityType {key: $key})
      OPTIONAL MATCH (et)-[:HAS_PROPERTY]->(p:PropertyDefinition)
      WHERE p.dataType <> 'document'
      RETURN collect(p.key) AS property_keys
      `,
      { key: entityTypeKey },
    );
    const record = result.records[0];
    return record === undefined ? [] : (record.get("property_keys") as string[]);
  });

  await validateExistingVectorIndexedProperties(
    driver,
    toPascalCase(entityTypeKey),
    entityTypeKey,
    propertyKeys,
  );
  await dropVectorIndex(driver, entityTypeKey);
  await createVectorIndex(driver, entityTypeKey, dimensions, propertyKeys);
}
